### PR TITLE
[Feature] Add callback for Text Generation Pipelines

### PR DESCRIPTION
**Pull request by rahul-tuli**
This PR adds a callback attribute to TextGenerationInput

`callback` is a callable that will be invoked on each generated token, and returns a boolean value.

If callback returns false, further generation of tokens is stopped

Test Code:

```python
from deepsparse import Pipeline

start = 0
def dummy_callback(token):
    global start
    start += 1
    return start < 3


if __name__ == "__main__":
    codegen_pipeline = Pipeline.create(
        task="text_generation",
        model_path = "zoo:nlg/text_generation/codegen_mono-350m/pytorch/huggingface/bigpython_bigquery_thepile/base-none"
    )
    
    inputs = {
        "sequences": ["def fib(a, b, accumulator=0)"]*3,
        "callback": dummy_callback,
        "return_logits": True,
    }
    
    outputs = codegen_pipeline(**inputs)
    assert ( shape:=outputs.logits.shape[1]) == 3, f"Expected 3 tokens, got {shape}"
```

Also added above logic in an automated test (This is not yet active on GHA, but tested locally)

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205276886236977

### CODE REVIEW

**Review by dbogunowicz — DISMISSED** [review]
LGTM, but still not sure how useful this function is. Was this a request from product/QA?

**Review by Satrat — DISMISSED** [review]

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 1]
```diff
@@ -457,6 +466,13 @@ def engine_forward(
                     ):
                         break
 
+                    if callback is not None and not callback(token):
+                        _LOGGER.info(
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:470]: This should be a debug log. Shouldn't log this on each inference

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 2]
```diff
@@ -91,6 +91,14 @@ class Config:
         "`streamer.put(token_ids)` and the streamer is responsible "
         "for any further processing.",
     )
+    callback: Optional[Callable[[Any], bool]] = Field(
+        default=None,
+        description="Callable that will be invoked "
+        "on each generated token. The callable must return a "
+        "Boolean value. If invocation returns `True`, the "
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:98]: Boolean value not required. Can return None for no effect

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 1 above)
Comment (rahul-tuli) [reply src/deepsparse/transformers/pipelines/text_generation.py:470]: No, this *only* logs when the callback returns False; which should be done right?

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 2 above)
Comment (rahul-tuli) [reply src/deepsparse/transformers/pipelines/text_generation.py:98]: Updated logic to only stop when `False` is returned

**Review by Satrat — APPROVED** [review]

**Review by bfineran — APPROVED** [review]

**Pull request merged**